### PR TITLE
Support for dev-mode-only conditional dependencies

### PR DESCRIPTION
**Pull request by aloubyansky**
This change introduces support for extension conditional dependencies that include dev mode as one of the conditions to enable a dependency.

The simplest form of a conditional dependency would be one extension declaring a dev mode conditional dependency on another extension w/o further conditions, i.e. dev mode will be the only condition. For example
```
            <plugin>
                <groupId>io.quarkus</groupId>
                <artifactId>quarkus-extension-maven-plugin</artifactId>
                <version>${quarkus.version}</version>
                <executions>
                    <execution>
                        <phase>compile</phase>
                        <goals>
                            <goal>extension-descriptor</goal>
                        </goals>
                        <configuration>
                            <deployment>${project.groupId}:${project.artifactId}-deployment:${project.version}</deployment>
                            <conditionalDevDependencies>
                                <extension>${project.groupId}:quarkus-blue:${project.version}</extension>
                            </conditionalDevDependencies>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```
An extension that is configured with the above `quarkus-extension-maven-plugin` configuration will declared a dev-mode-only dependency on extension `quarkus-blue`. `quarkus-blue` will be a regular extension, it will not have to configure anything on its side.

If an extension declaring a dev mode dependency on `quarkus-blue` also needs a compile time dependency on `quarkus-blue`, it can add an optional or provided dependency on`quarkus-blue`.

@phillip-kruger does this look Ok to you? I haven't added the docs yet, otherwise, it should be usable.

### CODE REVIEW

**Comment by phillip-kruger** [comment]
@aloubyansky can you show an example on how an extension will pull in it's own dev ui (runtime json-rpc services) ?

**Comment by aloubyansky** [comment]
Oh, sorry, I'm waking up :)
An extension may configure a dev mode dependency on another extension. Just a min I'll push an example.

**Comment by aloubyansky** [comment]
Here is a very simple example https://github.com/aloubyansky/playground/tree/dev-mode-conditional-deps

`quarkus-blue` is a regular extension project.
`quarkus-red` adds a dev mode only dependency on `quarkus-blue` in https://github.com/aloubyansky/playground/blob/dev-mode-conditional-deps/quarkus-red/runtime/pom.xml#L35
It doesn't add any other (Maven) dependency on it but it could, for example optional or provided, in case it needs it on the compile classpath.
An application adds a dependency on `quarkus-red` https://github.com/aloubyansky/playground/blob/dev-mode-conditional-deps/app/pom.xml#L16

The application can be launched in dev mode with `mvn clean compile quarkus:dev`

**Comment by maxandersen** [comment]
Nice. 

Haven't tried it yet but just curious - does the sbom generation captures this already or would that require more changes ?

- does it work with Gradle ?

- I'm going to assume it works with jbang as it would just be handled by Quarkus build time dep resolver?

- any downsides you spotted when implementing it ?

**Comment by aloubyansky** [comment]
Gradle will need a minor adjustment. Otherwise it should work.

**Comment by aloubyansky** [comment]
All done

**Comment by phillip-kruger** [comment]
@aloubyansky - I am still not sure how this should work for example moving dev-ui to a dev only dependency. For example, we currently have an extension that has a dev ui part, let's use arc extension to illustrate. So the current version of arc extension has some stuff in deployment for dev ui (Build items and Processor) and some stuff in runtime (JsonRPC Services). Should this be moved to a new extension (arc-dev-ui) ? And then the arc extension has a dev mode only dependency on this ? 

Currently, Dev UI used the extension as the base (and namespace) for everything it does. We don't want to list 2 extensions in Dev UI (one for arc and one for arc-dev-ui). So this will be a huge change on how dev ui works at the moment. 

Ideally, what we need is that the builditems still stay in arc deployment, and arc runtime has a dev mode only dependency on a jar (not an extension) that contains the json-rpc service. This way there is still only one extension, but the runtime dev ui classes is dev mode only. The deployment side does not matter as those are not included in the final artifact.

Or am I missing something ?

**Comment by aloubyansky** [comment]
> @aloubyansky - I am still not sure how this should work for example moving dev-ui to a dev only dependency. For example, we currently have an extension that has a dev ui part, let's use arc extension to illustrate. So the current version of arc extension has some stuff in deployment for dev ui (Build items and Processor) and some stuff in runtime (JsonRPC Services). Should this be moved to a new extension (arc-dev-ui) ? And then the arc extension has a dev mode only dependency on this ?

The way this PR implements it - yes, that would be the way.

> Currently, Dev UI used the extension as the base (and namespace) for everything it does. We don't want to list 2 extensions in Dev UI (one for arc and one for arc-dev-ui). So this will be a huge change on how dev ui works at the moment.

Ok, I see, I didn't think about it.

> Ideally, what we need is that the builditems still stay in arc deployment, with a dependency on a dev-mode-only-runtime (not an extension) that contains the json-rpc service, and still the normal runtime dependency that contains the normal runtime. This way there is still only one extension, but the runtime dev ui classes is dev mode only. The deployment side does not matter as those are not included in the final artifact.

I can add support for a non-extension conditional dependency.
I think, ideally, it'd be good to isolate (dependency-wise) even build steps and other dev-ui related resources from those used for the build. This dev-ui-only extension should be `unlisted: true` (in the metadata). Does DEV UI display unlisted extensions?

**Comment by phillip-kruger** [comment]
> I can add support for a non-extension conditional dependency.

Yes please. 

> I think, ideally, it'd be good to isolate (dependency-wise) even build steps and other dev-ui related resources from those used for the build. This dev-ui-only extension should be unlisted: true (in the metadata). Does DEV UI display unlisted extensions?

I am not sure why dependencies for dev ui needs to be isolated in the deployment module ? I think we need to be careful to not overuse extension. creating an extension just to show the dev ui for an extension, means that my extension is not really an extension. Let's keep extensions (conceptually) extensions.

**Comment by aloubyansky** [comment]
I pushed it to https://github.com/aloubyansky/quarkus/tree/conditional-non-extension-dep
It builds on top of this one but so far it works only for one Maven resolver. If you were to give it a try, you'd launch dev mode with `-Dquarkus.bootstrap.incubating-model-resolver=false`.
I'll look into how much work it'll take to support this in Gradle next.

**Comment by aloubyansky** [comment]
See an example in the description of this PR.

**Comment by phillip-kruger** [comment]
> See an example in the description of this PR.

Ok, so a normal dependency will work the same way as a dependency another extension ? Can I do this from the deployment module ?

**Comment by aloubyansky** [comment]
No, it'd be a runtime conditional dependency. Do you need a deployment-only conditional dependency?

**Comment by phillip-kruger** [comment]
No, i was just checking, so we add this to the runtime module. I'll give it a go tomorrow

File: docs/src/main/asciidoc/conditional-extension-dependencies.adoc [hunk 1]
```diff
@@ -9,21 +9,21 @@ include::_attributes.adoc[]
...
 
-The condition which triggers activation of an extension is configured in the extension's descriptor, which is included into the runtime artifact of the extension as `META-INF/quarkus-extension.properties`. Given that extension descriptor is generated by the Quarkus plugin at extension build time, extension developers can add the following configuration to express the condition which would have to be satisfied for the extension to be activated:
+The condition which triggers activation of an extension is configured in the extension's `META-INF/quarkus-extension.properties`, which is included in the runtime artifact of the extension.Extension developers can add the following configuration to express the condition which would have to be satisfied for the extension to be activated:
```
Comment (gsmet) [inline docs/src/main/asciidoc/conditional-extension-dependencies.adoc:26]:
```suggestion
The condition which triggers activation of an extension is configured in the extension's `META-INF/quarkus-extension.properties`, which is included in the runtime artifact of the extension. Extension developers can add the following configuration to express the condition which would have to be satisfied for the extension to be activated:
```

File: docs/src/main/asciidoc/conditional-extension-dependencies.adoc [hunk 2]
```diff
@@ -171,10 +173,61 @@ Conditional dependencies can also be configured in the Quarkus extension descrip
...
+          </execution>
+        </executions>
+      </plugin>
+
+  <!-- SKIPPED CONTENT -->
+----
+<1> the runtime Quarkus extension artifact ID;
+<2> the goal that generates the extension descriptor which every Quarkus runtime extension project should be configured with;
+<3> the Dev mode conditional dependency configuration element;
+<4> the artifact coordinates of conditional dependencies on extensions that should be evaluated only if an application is launched in Dev mode.
```
Comment (gsmet) [inline docs/src/main/asciidoc/conditional-extension-dependencies.adoc:227]: I think the convention is to use `dev mode` without the capitalized `Dev`. Should probably be adjusted everywhere.

**Review by gsmet — APPROVED** [review]
Looks good to me.
I spotted some minor doc things but I would recommend to merge this and fix them afterwards so that you can make progress without waiting on CI.

**Comment by geoand** [comment]
I added the `release/noteworthy-feature` label as this could be important for extensions.

cc @jmartisk @iocanel who I think have run into this in the past

**Pull request merged**